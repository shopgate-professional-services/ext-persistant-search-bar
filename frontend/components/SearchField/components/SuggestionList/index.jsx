import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import List from './components/List';
import SearchSuggestion from './components/SearchSuggestion';
import connect from './connector';
import styles from './style';

/**
 * The SuggestionList component.
 */
class SuggestionList extends Component {
  static propTypes = {
    bottomHeight: PropTypes.number.isRequired,
    isIOSTheme: PropTypes.func.isRequired,
    onClick: PropTypes.func.isRequired,
    fetching: PropTypes.bool,
    suggestions: PropTypes.arrayOf(PropTypes.string),
  }

  static defaultProps = {
    suggestions: [],
    fetching: false,
  }

  /**
   * @param { Object } nextProps Next props.
   * @return {boolean}
   */
  shouldComponentUpdate(nextProps) {
    return nextProps.fetching === false && nextProps.suggestions;
  }

  /**
   * @return {JSX}
   */
  render() {
    const { onClick, suggestions, bottomHeight } = this.props;

    if (!suggestions || suggestions.length === 0) {
      return null;
    }

    return (
      <div className={this.props.isIOSTheme() ?
        classnames(styles.iosList, styles.bottom(bottomHeight)) :
        classnames(styles.gmdList, styles.bottom(bottomHeight))}
      >
        <List>
          {suggestions.map(suggestion =>
        (<SearchSuggestion
          key={suggestion}
          suggestion={suggestion}
          onClick={e => onClick(e, suggestion)}
        />))}
        </List>
      </div>
    );
  }
}

export default connect(SuggestionList);
