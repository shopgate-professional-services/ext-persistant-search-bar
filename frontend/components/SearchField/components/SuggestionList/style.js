import { css } from 'glamor';
import { HEADER_HEIGHT, IOS_SEARCH_HEIGHT, GMD_SEARCH_HEIGHT } from '../../style';

/**
 * Manipulates the css when keyboard is opened to keep the UI scrollable.
 * @param {number} value Keyboard overlap value.
 * @returns {string}
 */
const bottom = value => css({
  paddingBottom: value,
}).toString();

const iosList = css({
  fontSize: 16,
  fontWeight: 400,
  bottom: 0,
  position: 'fixed',
  left: 0,
  right: 0,
  top: `calc(var(--safe-area-inset-top) + ${HEADER_HEIGHT}px + ${IOS_SEARCH_HEIGHT}px )`,
  backgroundColor: '#fff',
  overflowY: 'scroll',
  zIndex: 3,
  borderTop: '0.5px #cecece solid',
  paddingTop: 5,
}).toString();

const gmdList = css({
  fontSize: 16,
  fontWeight: 400,
  bottom: 0,
  position: 'fixed',
  left: 0,
  right: 0,
  top: `calc(var(--safe-area-inset-top) + ${HEADER_HEIGHT}px + ${GMD_SEARCH_HEIGHT}px )`,
  backgroundColor: '#fff',
  overflowY: 'scroll',
  zIndex: 3,
  borderTop: '0.5px #cecece solid',
  paddingTop: 5,
}).toString();

const item = css({
  alignItems: 'center',
  background: '#fff',
  display: 'flex',
  marginTop: 2,
  outline: 0,
  padding: '10px 16px 10px 50px',
  width: '100%',
  textAlign: 'left',
  borderBottom: '0.5px #cecece solid',
}).toString();

export default {
  bottom,
  iosList,
  gmdList,
  item,
};
