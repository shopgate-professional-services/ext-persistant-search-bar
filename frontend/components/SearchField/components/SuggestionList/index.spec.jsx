import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { createMockStore } from '@shopgate/pwa-common/store';
import SuggestionList from './index';

const store = createMockStore();

let mockedFetchingState;
jest.mock('@shopgate/pwa-common-commerce/search/selectors', () => ({
  getSuggestions: () => ([
    'foo',
    'foo bar',
    'foo bar buz',
    'foo bar buz quz',
  ]),
  getSuggestionsFetchingState: () => mockedFetchingState,
}));
const mockProps = {
  isIOSTheme: () => {},
};

describe('<SuggestionList />', () => {
  beforeEach(() => {
    mockedFetchingState = false;
  });

  it('should render a list of suggestions', () => {
    const wrapper = mount((
      <Provider store={store}>
        <SuggestionList bottomHeight={10} onClick={() => {}} {...mockProps} />
      </Provider>
    ));

    expect(wrapper).toMatchSnapshot();
  });
});
