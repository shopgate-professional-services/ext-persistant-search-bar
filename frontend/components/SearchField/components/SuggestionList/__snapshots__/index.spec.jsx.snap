// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuggestionList /> should render a list of suggestions 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(SuggestionList)
    bottomHeight={10}
    isIOSTheme={[Function]}
    onClick={[Function]}
  >
    <SuggestionList
      bottomHeight={10}
      dispatch={[Function]}
      fetching={false}
      isIOSTheme={[Function]}
      onClick={[Function]}
      suggestions={
        Array [
          "foo",
          "foo bar",
          "foo bar buz",
          "foo bar buz quz",
        ]
      }
    >
      <div
        className="css-rtftpe css-90ryaa"
      >
        <List>
          <List
            className=""
          >
            <ul
              className=""
            >
              <ListItem
                className="css-1brrprg css-1ym3l2n"
                isSelected={false}
                key="child-0/.$foo"
              >
                <li
                  className="css-1p5q8vq css-1brrprg css-1ym3l2n css-16gj3o1"
                  data-test-id="listItem"
                >
                  <div
                    className="css-1nnz83b"
                  >
                    <SearchSuggestion
                      key="foo"
                      onClick={[Function]}
                      suggestion="foo"
                    >
                      <Item
                        className={null}
                        image={null}
                        isDisabled={false}
                        isSelected={false}
                        link={null}
                        onClick={[Function]}
                        rightComponent={null}
                        testId={null}
                        title="foo"
                      >
                        <div
                          aria-hidden={true}
                          data-test-id={null}
                          onClick={[Function]}
                        >
                          <Glow
                            className={null}
                            color="#f7f7f7"
                            styles={
                              Object {
                                "container": null,
                                "glow": null,
                              }
                            }
                          >
                            <div
                              aria-hidden={true}
                              className="css-13azwyo"
                              onClick={[Function]}
                              style={null}
                            >
                              <div
                                className="css-1bmd5dz"
                                style={Object {}}
                              />
                              <Grid
                                className="css-1qcr7hz"
                                component="div"
                                wrap={false}
                              >
                                <div
                                  className="css-1qcr7hz css-h7r55a"
                                >
                                  <div
                                    className="css-jwxytc"
                                  />
                                  <GridItem
                                    className="css-10fxwvh"
                                    component="div"
                                    grow={1}
                                    shrink={1}
                                  >
                                    <div
                                      className="css-10fxwvh css-1gs0ko2"
                                    >
                                      foo
                                    </div>
                                  </GridItem>
                                </div>
                              </Grid>
                            </div>
                          </Glow>
                        </div>
                      </Item>
                    </SearchSuggestion>
                  </div>
                </li>
              </ListItem>
              <ListItem
                className="css-1brrprg css-1ym3l2n"
                isSelected={false}
                key="child-1/.$foo bar"
              >
                <li
                  className="css-1p5q8vq css-1brrprg css-1ym3l2n css-16gj3o1"
                  data-test-id="listItem"
                >
                  <div
                    className="css-1nnz83b"
                  >
                    <SearchSuggestion
                      key="foo bar"
                      onClick={[Function]}
                      suggestion="foo bar"
                    >
                      <Item
                        className={null}
                        image={null}
                        isDisabled={false}
                        isSelected={false}
                        link={null}
                        onClick={[Function]}
                        rightComponent={null}
                        testId={null}
                        title="foo bar"
                      >
                        <div
                          aria-hidden={true}
                          data-test-id={null}
                          onClick={[Function]}
                        >
                          <Glow
                            className={null}
                            color="#f7f7f7"
                            styles={
                              Object {
                                "container": null,
                                "glow": null,
                              }
                            }
                          >
                            <div
                              aria-hidden={true}
                              className="css-13azwyo"
                              onClick={[Function]}
                              style={null}
                            >
                              <div
                                className="css-1bmd5dz"
                                style={Object {}}
                              />
                              <Grid
                                className="css-1qcr7hz"
                                component="div"
                                wrap={false}
                              >
                                <div
                                  className="css-1qcr7hz css-h7r55a"
                                >
                                  <div
                                    className="css-jwxytc"
                                  />
                                  <GridItem
                                    className="css-10fxwvh"
                                    component="div"
                                    grow={1}
                                    shrink={1}
                                  >
                                    <div
                                      className="css-10fxwvh css-1gs0ko2"
                                    >
                                      foo bar
                                    </div>
                                  </GridItem>
                                </div>
                              </Grid>
                            </div>
                          </Glow>
                        </div>
                      </Item>
                    </SearchSuggestion>
                  </div>
                </li>
              </ListItem>
              <ListItem
                className="css-1brrprg css-1ym3l2n"
                isSelected={false}
                key="child-2/.$foo bar buz"
              >
                <li
                  className="css-1p5q8vq css-1brrprg css-1ym3l2n css-16gj3o1"
                  data-test-id="listItem"
                >
                  <div
                    className="css-1nnz83b"
                  >
                    <SearchSuggestion
                      key="foo bar buz"
                      onClick={[Function]}
                      suggestion="foo bar buz"
                    >
                      <Item
                        className={null}
                        image={null}
                        isDisabled={false}
                        isSelected={false}
                        link={null}
                        onClick={[Function]}
                        rightComponent={null}
                        testId={null}
                        title="foo bar buz"
                      >
                        <div
                          aria-hidden={true}
                          data-test-id={null}
                          onClick={[Function]}
                        >
                          <Glow
                            className={null}
                            color="#f7f7f7"
                            styles={
                              Object {
                                "container": null,
                                "glow": null,
                              }
                            }
                          >
                            <div
                              aria-hidden={true}
                              className="css-13azwyo"
                              onClick={[Function]}
                              style={null}
                            >
                              <div
                                className="css-1bmd5dz"
                                style={Object {}}
                              />
                              <Grid
                                className="css-1qcr7hz"
                                component="div"
                                wrap={false}
                              >
                                <div
                                  className="css-1qcr7hz css-h7r55a"
                                >
                                  <div
                                    className="css-jwxytc"
                                  />
                                  <GridItem
                                    className="css-10fxwvh"
                                    component="div"
                                    grow={1}
                                    shrink={1}
                                  >
                                    <div
                                      className="css-10fxwvh css-1gs0ko2"
                                    >
                                      foo bar buz
                                    </div>
                                  </GridItem>
                                </div>
                              </Grid>
                            </div>
                          </Glow>
                        </div>
                      </Item>
                    </SearchSuggestion>
                  </div>
                </li>
              </ListItem>
              <ListItem
                className="css-1brrprg"
                isSelected={false}
                key="child-3/.$foo bar buz quz"
              >
                <li
                  className="css-1p5q8vq css-1brrprg css-16gj3o1"
                  data-test-id="listItem"
                >
                  <div
                    className="css-1nnz83b"
                  >
                    <SearchSuggestion
                      key="foo bar buz quz"
                      onClick={[Function]}
                      suggestion="foo bar buz quz"
                    >
                      <Item
                        className={null}
                        image={null}
                        isDisabled={false}
                        isSelected={false}
                        link={null}
                        onClick={[Function]}
                        rightComponent={null}
                        testId={null}
                        title="foo bar buz quz"
                      >
                        <div
                          aria-hidden={true}
                          data-test-id={null}
                          onClick={[Function]}
                        >
                          <Glow
                            className={null}
                            color="#f7f7f7"
                            styles={
                              Object {
                                "container": null,
                                "glow": null,
                              }
                            }
                          >
                            <div
                              aria-hidden={true}
                              className="css-13azwyo"
                              onClick={[Function]}
                              style={null}
                            >
                              <div
                                className="css-1bmd5dz"
                                style={Object {}}
                              />
                              <Grid
                                className="css-1qcr7hz"
                                component="div"
                                wrap={false}
                              >
                                <div
                                  className="css-1qcr7hz css-h7r55a"
                                >
                                  <div
                                    className="css-jwxytc"
                                  />
                                  <GridItem
                                    className="css-10fxwvh"
                                    component="div"
                                    grow={1}
                                    shrink={1}
                                  >
                                    <div
                                      className="css-10fxwvh css-1gs0ko2"
                                    >
                                      foo bar buz quz
                                    </div>
                                  </GridItem>
                                </div>
                              </Grid>
                            </div>
                          </Glow>
                        </div>
                      </Item>
                    </SearchSuggestion>
                  </div>
                </li>
              </ListItem>
            </ul>
          </List>
        </List>
      </div>
    </SuggestionList>
  </Connect(SuggestionList)>
</Provider>
`;
